File: articles/calculus-for-systems-modeling/articles/elasticity-sensitivity-and-marginal-response/r/elasticity_sensitivity_audit.R
response_function <- function(x) {
  10 * sqrt(x + 1)
}

analytic_derivative <- function(x) {
  5 / sqrt(x + 1)
}

finite_difference_derivative <- function(x, h = 1e-5) {
  (response_function(x + h) - response_function(x - h)) / (2 * h)
}

classify_response <- function(elasticity) {
  if (is.na(elasticity)) {
    "elasticity undefined"
  } else if (abs(elasticity) < 1) {
    "inelastic local response"
  } else if (abs(elasticity) == 1) {
    "unit elastic local response"
  } else {
    "elastic local response"
  }
}

audit_point <- function(x) {
  y <- response_function(x)
  derivative <- analytic_derivative(x)
  finite_difference <- finite_difference_derivative(x)
  error <- abs(derivative - finite_difference)

  warnings <- c()
  elasticity <- NA_real_

  if (x == 0) {
    warnings <- c(warnings, "input is zero; proportional input change requires care")
  }

  if (y == 0) {
    warnings <- c(warnings, "output is zero; elasticity undefined")
  }

  if (x != 0 && y != 0) {
    elasticity <- (x / y) * derivative
  }

  if (error > 1e-5) {
    warnings <- c(warnings, "finite-difference check differs from analytic derivative")
  }

  if (x < 0.1) {
    warnings <- c(warnings, "near-zero baseline; normalize with caution")
  }

  data.frame(
    x = x,
    value = y,
    derivative = derivative,
    elasticity = elasticity,
    finite_difference_derivative = finite_difference,
    absolute_error = error,
    response_class = classify_response(elasticity),
    warning = paste(warnings, collapse = "; ")
  )
}

results <- do.call(rbind, lapply(c(0, 0.5, 1, 4, 9, 24), audit_point))
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_elasticity_sensitivity_audit.csv", row.names = FALSE)
print(results)

File: articles/calculus-for-systems-modeling/articles/elasticity-sensitivity-and-marginal-response/julia/elasticity_sensitivity_audit.jl
response_function(x) = 10.0 * sqrt(x + 1.0)
analytic_derivative(x) = 5.0 / sqrt(x + 1.0)
finite_difference_derivative(x; h=1e-5) = (response_function(x+h) - response_function(x-h)) / (2h)

function classify_response(e)
    if isnothing(e)
        return "elasticity undefined"
    elseif abs(e) < 1.0
        return "inelastic local response"
    elseif abs(e) == 1.0
        return "unit elastic local response"
    else
        return "elastic local response"
    end
end

println("x,value,derivative,elasticity,finite_difference_derivative,absolute_error,response_class,warning")
for x in [0.0, 0.5, 1.0, 4.0, 9.0, 24.0]
    y = response_function(x)
    d = analytic_derivative(x)
    fd = finite_difference_derivative(x)
    err = abs(d - fd)
    e = (x == 0.0 || y == 0.0) ? nothing : (x / y) * d
    warning = x == 0.0 ? "input is zero; proportional input change requires care" : ""
    eout = isnothing(e) ? "NA" : string(e)
    println("$x,$y,$d,$eout,$fd,$err,$(classify_response(e)),$warning")
end
